sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -189,16 +189,19 @@ object SubExprUtils extends PredicateHelper {
    */
   def getOuterReferences(expr: Expression): Seq[Expression] = {
     val outerExpressions = ArrayBuffer.empty[Expression]
-    expr transformDown {
```

Comment:
We don't use the transformed expression, so should better use `foreach`

Reply:
The test failures are related to this change?

Reply:
This is tricky. I tried updating this function as well and discovered that `transformDown` is actually needed. 
https://github.com/apache/spark/blob/8fa739fb9d3bd49efa7df7525df18b7111b0131e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala#L193-L198
Notice the `newExpr` here has outer references removed. Say we have an expression `min(outer(a))` and if we use `foreach`, we will collect both `min(a)` and `a`, but if we use `transformDown`, the expression tree `min(outer(a))` will be transformed into `min(a)` so no more outer references will be collected.
(I added the comment but it's probably not very clear.)

Reply:
I updated it to write a manual tree travesal.